src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -327,6 +410,148 @@ public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeader
+
+            // C's chain remains at same height after A presents the invalid header
+            Assert.True(chainedHeaderTree.GetPeerTipChainedHeaderByPeerId(2).Height == peerThreeTip.Height);
+        }
+
+        /// <summary>
+        /// Issue 9 @ Create chained header tree component #1321
+        /// Check that everything is always consumed.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_MultiplePeers_CheckEverythingIsConsumed()
+        {
+            // Chain A is presented by default peer:
+            // 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            TestContext testContext = new TestContextBuilder().Build();
+            ChainedHeaderTree chainedHeaderTree = testContext.ChainedHeaderTree;
+            ChainedHeader chainTip = testContext.ExtendAChain(7);
+            chainedHeaderTree.Initialize(chainTip, true);
+
+            // Chain A is extended by Peer 1:
+            // 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a - 9a - 10a - 11a - 12a - 13a - 14a
+            ChainedHeader peer1ChainTip = testContext.ExtendAChain(6, chainTip);
+            List<BlockHeader> listOfPeer1NewHeaders = testContext.ChainedHeaderToList(peer1ChainTip, 6);
+            List<BlockHeader> peer1NewHeaders = listOfPeer1NewHeaders.GetRange(0, 4);
+
+            // 4 headers are processed 2 are unprocessed.
+            ConnectNewHeadersResult connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, peer1NewHeaders);
+            Assert.Equal(2, peer1ChainTip.Height - connectedNewHeadersResult.Consumed.Height);
+
+            // Remaining headers are presented.
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, listOfPeer1NewHeaders.GetRange(4, 2));
+            
+            // All headers are now consumed.
+            Assert.Equal(connectedNewHeadersResult.Consumed.HashBlock, peer1ChainTip.HashBlock);
+
+            // Peer 2 presents headers that that are already in the tree: 11a - 12a - 13a - 14a
+            // As well as some that are not in it yet: 15a - 16a
+            ChainedHeader peer2ChainTip = testContext.ExtendAChain(2, peer1ChainTip); 
+            List<BlockHeader> listOfNewAndExistingHeaders = testContext.ChainedHeaderToList(peer2ChainTip, 6);
+
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(2, listOfNewAndExistingHeaders);
+            Assert.True(connectedNewHeadersResult.Consumed.Height == peer2ChainTip.Height);
+
+            // Peer 3 presents a list of headers that are already in the tree
+            // and the rest of the headers from a new fork to the tree.
+            // 13a - 14a - 15a - 16a
+            //     - 14b - 15b - 16b
+            ChainedHeader peer3ChainTip = testContext.ExtendAChain(2, peer2ChainTip.Previous.Previous.Previous);
+            List<BlockHeader> listOfPeer3Headers = testContext.ChainedHeaderToList(peer3ChainTip, 5);
```

Comment:
Here the comment does not correspond with the code is your extending the chain only by two and also you're taking five blocks while in comment you have extension of three and you're taking four

Reply:
Fixed and changed all the comments so they follow the h1=h2=h3 format where these refer to heights